util/unresolved_addr.go

```diff
@@ -21,22 +21,38 @@ import (
 	"net"
 )
 
-// MakeUnresolvedAddr creates a new UnresolvedAddr from a network
-// and raw address string.
-func MakeUnresolvedAddr(network string, str string) UnresolvedAddr {
+// MakeUnresolvedAddr populates an UnresolvedAddr from a network and raw
+// address string.
+func MakeUnresolvedAddr(network, addr string) UnresolvedAddr {
 	return UnresolvedAddr{
 		NetworkField: network,
-		AddressField: str,
+		AddressField: addr,
 	}
 }
 
+// NewUnresolvedAddr creates a new UnresolvedAddr from a network and raw
+// address string.
+func NewUnresolvedAddr(network, addr string) *UnresolvedAddr {
+	return &UnresolvedAddr{
+		NetworkField: network,
+		AddressField: addr,
+	}
+}
+
+// Note that we make *UnresolvedAddr implement the net.Addr interface, not
+// UnresolvedAddr. This is done because assigning a non-empty struct to an
+// interface requires an allocation, while assigning a pointer to an interface
+// is allocation free. Using an *UnresolvedAddr makes it both clear that an
+// allocation is occurring and allows us to avoid an allocation when an
+// UnresolvedAddr is a field of a struct (e.g. NodeDescriptor.Address).
+
```

Comment:
put a `var _ net.Addr = &UnresolvedAddr{}` here so that the comment above is anchored